py/src/nupic/optimization/ae.py

```diff
@@ -0,0 +1,898 @@
+#!/usr/bin/python3
+"""
+Automatic Experimenter
+Written by David McDougall, 2018-2019
+
+This is a parameter optimization framework.
+ * It methodically records the results of different sets of parameters and
+analyses the results.  It then automatically suggests and evaluates
+modifications to the parameters.
+ * It exposes a convenient API for users to hook their program into this.
+ * The framework allows for testing each set of parameters several times and
+calculates the average and standard deviation of the results.  It also
+calculates the confidence that a parameter change caused the score to change.
+ * It is extensible: new methods for automated parameter optimization can be
+added.  Currently this implements a basic grid search strategy.  In the future I
+hope to add a particle swarming method.
+
+To use this module, structure experiments as follows:
+    ExperimentModule is a python3 module containing the model to be optimized as
+    well as code to evaluate model performance.
+
+    ExperimentModule.default_parameters = {}
+    This global dictionary contains all of the parameters to modify.
+    Parameters must be one of the following types: dict, tuple, float, int.
+    Parameters can be nested in multiple levels of dictionaries and tuples.
+
+    ExperimentModule.main(parameters=default_parameters, argv=None, verbose=True)
+    Returns (float) performance of parameters, to be maximized.
+
+Usage: $ ae.py [ae-arguments] ExperimentModule.py [experiment-arguments]
```

Comment:
can you provide a runnable demo example? That would be worth 1000 words for first-timers wishing to use the code. 
MNIST would be an ideal candidate and you've already optimized that code for me once, so you should have the "experiment" files ready.